### PR TITLE
Gmerz/config refactor

### PR DESCRIPTION
**Pull request by grantmerz**
## Problem & Solution Description (including issue #)

This updates the config usage to reflect the changes in deepdisc.  Also cleans a few repeating lines and changes some StageConfig parameters 

## Code Quality
- [ ] My code follows [the code style of this project](https://rail-hub.readthedocs.io/en/latest/source/contributing.html#naming-conventions)
- [ ] I have written unit tests or justified all instances of `#pragma: no cover`; in the case of a bugfix, a new test that breaks as a result of the bug has been added
- [ ] My code contains relevant comments and necessary documentation for future maintainers; the change is reflected in applicable demos/tutorials (with output cleared!) and added/updated docstrings use the [NumPy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html)
- [ ] Any breaking changes, described above, are accompanied by backwards compatibility and deprecation warnings

### CODE REVIEW

File: src/rail/estimation/algos/deepdisc.py [hunk 1]
```diff
@@ -72,84 +69,65 @@ def train(config, all_metadata, train_head=True):
...
-        trainer.train(0, head_iters)
+        trainer.train(0, e1)
 
         if comm.is_main_process():
-            np.save(output_dir + output_name + "_losses", trainer.lossList)
-            # np.save(output_dir + output_name + "_val_losses", trainer.vallossList)
+            np.save(output_dir + run_name + "_losses", trainer.lossList)
+            # np.save(output_dir + run_name + "_val_losses", trainer.vallossList)
 
         return model
```
Comment (drewoldag) [inline src/rail/estimation/algos/deepdisc.py:105]: Can we go ahead and remove this return statement, and the return statement in the `else` condition?

File: tests/configs/COCO/cascade_mask_rcnn_mvitv2_b_in21k_100ep.py
Comment (drewoldag) [inline tests/configs/COCO/cascade_mask_rcnn_mvitv2_b_in21k_100ep.py:1]: Did you mean to move the `configs` directory into `tests`? I don't immediately understand the reasoning here.

File: src/rail/estimation/algos/deepdisc.py [hunk 2]
```diff
@@ -23,37 +23,33 @@
...
-    cfg_loader = get_loader_config(output_dir, batch_size)
 
-    e1 = epochs * 15
-    e2 = epochs * 10
-    e3 = epochs * 20
-    efinal = epochs * 35
+    e1 = epoch * head_epochs
+    e2 = epoch * 10
+    e3 = epoch * 20
+    efinal = epoch * full_epochs
```
Comment (drewoldag) [inline src/rail/estimation/algos/deepdisc.py:50]: minor point (not necessary to change in this PR) - do we really need to define these (`e1`, `e2`, `e3`, `efinal`)? Or can we use the `epoch * X` directly?

File: src/rail/estimation/algos/deepdisc.py [hunk 3]
```diff
@@ -72,86 +68,66 @@ def train(config, all_metadata, train_head=True):
...
         trainer.set_period(print_frequency)
 
-        trainer.train(0, head_iters)
+        trainer.train(0, e1)
 
         if comm.is_main_process():
-            np.save(output_dir + output_name + "_losses", trainer.lossList)
-            # np.save(output_dir + output_name + "_val_losses", trainer.vallossList)
+            np.save(output_dir + run_name + "_losses", trainer.lossList)
+            # np.save(output_dir + run_name + "_val_losses", trainer.vallossList)
```
Comment (drewoldag) [inline src/rail/estimation/algos/deepdisc.py:102]: Will this need to be uncommented and conditionally executed when we make `lossHook` conditional?

File: src/rail/estimation/algos/deepdisc.py [hunk 4]
```diff
@@ -72,86 +68,66 @@ def train(config, all_metadata, train_head=True):
...
+            losses = np.load(output_dir + run_name + "_losses.npy")
             losses = np.concatenate((losses, trainer.lossList))
-            np.save(output_dir + output_name + "_losses", losses)
+            np.save(output_dir + run_name + "_losses", losses)
 
-            # vallosses = np.load(output_dir + output_name + "_val_losses.npy")
+            # vallosses = np.load(output_dir + run_name + "_val_losses.npy")
             # vallosses = np.concatenate((vallosses, trainer.vallossList))
-            # np.save(output_dir + output_name + "_val_losses", vallosses)
+            # np.save(output_dir + run_name + "_val_losses", vallosses)
```
Comment (drewoldag) [inline src/rail/estimation/algos/deepdisc.py:128]: Will these lines need to be uncommented and conditionally executed when we make lossHook conditional?

File: src/rail/estimation/algos/deepdisc.py [hunk 5]
```diff
@@ -274,7 +249,7 @@ class DeepDiscEstimator(CatEstimator):
         numclasses=Param(int, 1, required=False, msg="The number of classes in the model."),
         epochs=Param(int, 20, required=False, msg="How many epochs to run estimation."),
         output_dir=Param(str, "./", required=False, msg="The directory to write output to."),
-        output_name=Param(str, "deepdisc_informer", required=False, msg="What to call the generated output."),
+        run_name=Param(str, "run", required=False, msg="Name of the training run."),
```
Comment (drewoldag) [inline src/rail/estimation/algos/deepdisc.py:252]: Good catch 😉

**Review by drewoldag — APPROVED** [review]
Generally this looks fine to me. I haven't had the opportunity to run the rail_wrap_test notebook, which seems to be our best end-to-end test, but assuming that you _have_ run it, this seems fine. 

My earlier question about the configs directory wasn't so much about whether it should be in the tests directory or somewhere else, but whether so many configs should be included in the package in general. If all the configs belong here, then as part of the documentation, it might be nice to have a manifest of them with some explanation of when to use them.  

@grantmerz you'll know best if the users of this package will already have the knowledge about how to use the configs and what the appropriate amount of documentation will be.

**Comment by drewoldag** [comment]
Also, I think the unit tests are failing now because there are no unit tests. The placeholder test was removed, and so the github workflow has nothing to execute. This isn't a major problem, but just in case you were curious why the unit tests are failing. 

It'll be better (for a lot of reasons) once we add some more unit tests.

**Comment by grantmerz** [comment]
> Generally this looks fine to me. I haven't had the opportunity to run the rail_wrap_test notebook, which seems to be our best end-to-end test, but assuming that you _have_ run it, this seems fine.
> 
> My earlier question about the configs directory wasn't so much about whether it should be in the tests directory or somewhere else, but whether so many configs should be included in the package in general. If all the configs belong here, then as part of the documentation, it might be nice to have a manifest of them with some explanation of when to use them.
> 
> @grantmerz you'll know best if the users of this package will already have the knowledge about how to use the configs and what the appropriate amount of documentation will be.

I included all configs because I have not yet gone through all of the config dependencies and determined what files are absolutely necessary.  The current config files are all under the detectron2 repo [here](https://github.com/facebookresearch/detectron2/tree/main/projects/ViTDet/configs), but not under the source directory, so I don't think it's easy to import them.  I think I would prefer to not include them in rail_deepdisc, but I'm not sure of the best way to point to them under the detectron2 repo and make sure a user has that directory

**Comment by drewoldag** [comment]
@grantmerz I see what you're getting at with the configs now. I would suggest that the path forward would be to create an issue for the future to determine if there is a better way to handle or import the configs. No reason for the configs to hold up this PR :)

**Pull request merged**